### ISSUE TITLE
private _ctypes unimportable on linux bundle

### ISSUE THREAD

**Issue by tlambert03** [beeware/briefcase#407]
Running into an issue with our linux build, and wondering if you might have any thoughts?

I built the AppImage with github actions running ubuntu-16.04 as described in the tutorial (full build log and artifacts [here](https://github.com/tlambert03/napari/runs/704370988?check_suite_focus=true)).  Building and packaging completed, but the packaging step did show a patchelf error:
```
Setting rpath in ELF file /home/runner/work/napari/napari/linux/napari/napari.AppDir/usr/lib/python3.8/config-3.8-x86_64-linux-gnu/python.o to $ORIGIN 
ERROR: Call to patchelf failed: 
wrong ELF type
 
ERROR: Failed to set rpath in ELF file: /home/runner/work/napari/napari/linux/napari/napari.AppDir/usr/lib/python3.8/config-3.8-x86_64-linux-gnu/python.o 
```

When I then downloaded the artifact and tried to run on a local ubuntu machine (18.04) it failed to import numpy with:
```
 File "/tmp/.mount_napari6jrUf5/usr/app_packages/vispy/util/logs.py", line 14, in <module>
    import numpy as np
  File "/tmp/.mount_napari6jrUf5/usr/app_packages/numpy/__init__.py", line 142, in <module>
    from . import core
  File "/tmp/.mount_napari6jrUf5/usr/app_packages/numpy/core/__init__.py", line 102, in <module>
    from . import _dtype_ctypes
  File "/tmp/.mount_napari6jrUf5/usr/app_packages/numpy/core/_dtype_ctypes.py", line 25, in <module>
    import _ctypes
ModuleNotFoundError: No module named '_ctypes'
```

any ideas about that one?  thanks!

**Comment by freakboy3742** [comment]
Can say I do, unfortunately... It *looks* like either the ctypes module hasn't been build correctly in the Linux support package, or there's something going awry with the packaging of the ctypes library. 

The first thing to check will be whether ctypes works *at all* using the support build. If you were to download the support package into an Ubuntu 16.04 docker image, start the packaged python installer, and try to run `import ctypes`, that will give you some initial indication.

**Comment by GenevieveBuckley** [comment]
The [Python-Apple-support)](https://github.com/beeware/Python-Apple-support) has quite a lot of patching or special handling involving `ctype`, fyi. The linux support package definitely does not. I couldn't find a repo matching "Python-Windows-Support", so I'm not sure what happens to `ctypes` there.

**Comment by freakboy3742** [comment]
@GenevieveBuckley On Windows, we actually use the officially distribute "Python embedded" zip file available from Python.org. 

The Apple builds are patched to make all the libraries statically linked; this is a requirement for iOS [1], but it doesn't *hurt* on a standalone macOS build, so we keep the same patch.

The Linux build should be pretty much "CPython source code as shipped"; however, taking a closer look, it appears that libffi (the major pre-requisite of ctypes) isn't *explicitly* being added to the build container, and as a result, it's possible ctypes hasn't been built correctly. I'll need to investigate further (or, if someone else wants to... don't let me stop them :-) )

---
[1] ... or, at least, it was prior to iOS 8, and I haven't had a chance to fix the underlying problem that would allow dynamic linking.

**Comment by tlambert03** [comment]
I can confirm that the hello world example in the tutorial doesn't work if you add `import ctypes`:
https://github.com/tlambert03/brieftest/runs/706716843?check_suite_focus=true#step:5:417

I'm afraid my docker-fu isn't quite up to snuff, but I directly downloaded the support package and ran `python3 -c "import ctypes"` and it also failed with `ModuleNotFoundError: No module named '_ctypes'`
https://github.com/tlambert03/brieftest/runs/706757611?check_suite_focus=true#step:5:88

**Comment by freakboy3742** [comment]
It appears that @tlambert03's analysis was correct; the CPython build didn't include libFFI, and thus didn't have a successful _ctypes build. 

I've just published updated support packages 3.8-b2, 3.7-b2, 3.6-b2 and 3.5-b2; these now include ctypes, and also do a more effective stripping of binary content (so the final AppImage binaries are now much smaller).

**Comment by tlambert03** [comment]
Thanks @freakboy3742!
